# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
part = pd.DataFrame(data.groupby('children')['debt'].agg(['mean']).sort_values(by='mean', ascending = False))
num = pd.DataFrame(data.groupby('children')['debt'].agg(['count']).sort_values(by='count', ascending = False))
debt = pd.DataFrame(data.groupby('children')['debt'].agg(['sum']).sort_values(by='sum', ascending = False))
part=part.merge(num, on='children', how='left')
part=part.merge(debt, on='children', how='right')
part = part.rename(columns={
'mean'
: 
'Доля невозврата'
, 
'count'
: 
'Количество заемщиков',
'sum'
: 
'Количество должников'
})
part

Доля невозврата  Количество заемщиков  Количество должников
children                                                             
0                0.075438                 14091                  1063
1                0.092346                  4808                   444
2                0.094542                  2052                   194
3                0.081818                   330                    27
4                0.097561                    41                     4
5                0.000000                     9                     0

**Вывод: Наибольшее количество задолженников присутствует в группе людей, имеющих четверых детей, наименьшее - среди людей, не имеющих детей. Для большинства групп (людей, имеющих одного ребенка, не имеющих детей и имещих 4 и 2 детей) прослеживается следующая тенденция: чем больше у человека детей, тем скорее он будет задолженником. Эта теория не распространяется на людей, имеющих 3 детей (среди них задолженников меньше, чем у отцов/матерей двух детей и одного ребенка).
Важно заметить, что точных выводов в данном случае мы получить не можем, так как категория людей, имеющих пятеро детей, слишком малочисленна для построения тенденции и учета полученных данных в общем выводе (5 клиентов в выборке).**  

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
part = pd.DataFrame(data.groupby('family_status')['debt'].agg(['mean']).sort_values(by='mean', ascending = False))
num = pd.DataFrame(data.groupby('family_status')['debt'].agg(['count']).sort_values(by='count', ascending = False))
debt = pd.DataFrame(data.groupby('family_status')['debt'].agg(['sum']).sort_values(by='sum', ascending = False))
part=part.merge(num, on='family_status', how='left')
part=part.merge(debt, on='family_status', how='right')
part = part.rename(columns={
'mean'
: 
'Доля невозврата'
, 
'count'
: 
'Количество заемщиков',
'sum'
: 
'Количество должников'
})
part

Доля невозврата  Количество заемщиков  \
family_status                                                  
женат / замужем               0.075606                 12261   
гражданский брак              0.093130                  4134   
Не женат / не замужем         0.097639                  2796   
в разводе                     0.070648                  1189   
вдовец / вдова                0.066246                   951   

                       Количество должников  
family_status                                
женат / замужем                         927  
гражданский брак                        385  
Не женат / не замужем                   273  
в разводе                                84  
вдовец / вдова                           63

**Вывод: Больше всего задолженностей имеется у людей, никогда не сосотоящих в браке, меньше всего - у людей, которые больше не состоят в браке (в любом виде) по каким бы то ни было причинам (они находятся в разводе или стали вдовой/вдовцом).** 

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
part = pd.DataFrame(data.groupby('total_income_category')['debt'].agg(['mean']).sort_values(by='mean', ascending = False))
num = pd.DataFrame(data.groupby('total_income_category')['debt'].agg(['count']).sort_values(by='count', ascending = False))
debt = pd.DataFrame(data.groupby('total_income_category')['debt'].agg(['sum']).sort_values(by='sum', ascending = False))
part=part.merge(num, on='total_income_category', how='left')
part=part.merge(debt, on='total_income_category', how='right')
part = part.rename(columns={
'mean'
: 
'Доля невозврата'
, 
'count'
: 
'Количество заемщиков',
'sum'
: 
'Количество должников'
})
part = part.sort_values(by='Доля невозврата', ascending = False)
part

Доля невозврата  Количество заемщиков  \
total_income_category                                          
E                             0.090909                    22   
C                             0.084982                 15921   
A                             0.080000                    25   
B                             0.070602                  5014   
D                             0.060172                   349   

                       Количество должников  
total_income_category                        
E                                         2  
C                                      1353  
A                                         2  
B                                       354  
D                                        21

**Вывод: Меньше всего задолженностей у людей с уровнем заработка от 30001 рублей до 50 тыс. рублей. Больше всего задолженностей среди людей с наименьшим уровнем заработка среди всех категорий (от 0 до 30 тыс. рублей). Для остальных трех категорий людей с наибольшим уровнем заработка (A, B, C) тенденция отсутствует - хотя среди людей с уровнем заработка от 200001 до 1000000 рублей задолженников меньше, чем среди людей с уровнем заработка больше 1000001, среди людей с уровнем заработка от 50001 до 200000 рублей задолженников больше, чем в обеих вышеперечисленных категориях(A, B).** 

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
part = pd.DataFrame(data.groupby('purpose_category')['debt'].agg(['mean']).sort_values(by='mean', ascending = False))
num = pd.DataFrame(data.groupby('purpose_category')['debt'].agg(['count']).sort_values(by='count', ascending = False))
debt = pd.DataFrame(data.groupby('purpose_category')['debt'].agg(['sum']).sort_values(by='sum', ascending = False))
part=part.merge(num, on='purpose_category', how='left')
part=part.merge(debt, on='purpose_category', how='right')
part = part.rename(columns={
'mean'
: 
'Доля невозврата'
, 
'count'
: 
'Количество заемщиков',
'sum'
: 
'Количество должников'
})
part

Доля невозврата  Количество заемщиков  \
purpose_category                                                  
операции с недвижимостью         0.072551                 10751   
операции с автомобилем           0.093480                  4279   
получение образования            0.092528                  3988   
проведение свадьбы               0.079118                  2313   

                          Количество должников  
purpose_category                                
операции с недвижимостью                   780  
операции с автомобилем                     400  
получение образования                      369  
проведение свадьбы                         183

**Вывод: наибольшее количество должников наблюдается среди тех людей, которые брали кредиты на покупку автомобиля или для получения образования, наименьшее - среди тех, кто воспользовался кредитом при проведении операций с недвижимостью или для свадьбы. Можно предположить, что в данном случае доля задолженников в каждой категории объясняется психологией клиента: люди чаще долго готовятся и отвественно подходят к таким событиям, как свадьба или покупка недвижимости, нежели получение образования или покупка автомобиля.**

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ: Пропуски в данных встречаются в столбцах, содержащих информацию о доходе и общем трудовом стаже заемщика. Это может быть объяснено двумя причинами: 1) Заемщики неофициально трудоустроены, не имеют постоянного заработка или по каким-то другим причинам не знают точной информации об общем стаже и регулярном доходе. 2)Потеря данных по техническим причинам* 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ: Медиана является наиболее объективным показателем для таких количественных значений, как зарплата или общее время трудового стажа в днях. Медиана показывает актуальное среднее значение для неоднородных данных. Другой показатель среднего значения - среднее аримфметическое - варьируется в зависимости от аномальных значений, составляющих меньшинство данных.* 

### Шаг 4: общий вывод.

Постараемся определить основные параметры клиентов, которые с большей вероятностью не будут должниками.
- Целевой аудиторией исследуемого продукта являются люди, не имеющие детей, среди этих клиентов также доля должников имеет наименьший показатель - 0.075438 (самая малая доля должников приходится на клиентов, имеющих пятерых детей, но таких всего 9 человек, что нельзя считать достаточно репрезентативным). 
- Хотя наименьшее количество должников приходится на клиентов, которые ранее состояли в браке, но уже не состоят в нем (являются вдовами/вдовцами или находятся в разводе), гораздо большее количество людей среди заемщиков женаты, доля должников среди них тоже невелика (находится на третьем месте в списке возрастания долей должников), так что последних можно считать целевой аудиторией продукта.
- Больше всего клиентов банка зарабатывают либо от 50001 до 200000 рублей (15921), либо от 200001 до 1000000 рублей (5014). Среди первых доля должников составляет 0.084982, что представляет собой большое значение относительно других категорий заработка (находится на втором месте в списке убывания долей должников). Доля задолженностей среди людей с более высоким уровнем заработка (от 200001 до 1000000 рублей) составляет 0.070602, являясь второй по счету наименьшей долей должников в рассматриваемой выборке (по уровню заработной платы). Таким образом, кажется целесообразным в работе обратить внимание на клиентов, попадающих в категорию С по уровню заработка. 
- Наименьшее количество должников в соответствии с целями взятия кредита приходится на самую большую по численности клиентов категорию - людей, которые берут кредиты для совершения операций с недвижимостью. Очевидно, именно они представляют собой ЦА продукта. 
